pkg/tsdb/elasticsearch/response_parser.go

```diff
@@ -753,3 +891,79 @@ func getErrorFromElasticResponse(response *es.SearchResponse) string {
 
 	return errorString
 }
+
+// flatten flattens multi-level objects to single level objects. It uses dot notation to join keys.
+func flatten(target map[string]interface{}) map[string]interface{} {
+	// On frontend maxDepth wasn't used but as we are processing on backend
+	// let's put a limit to avoid infinite loop. 10 was chosen arbitrary.
+	maxDepth := 10
+	currentDepth := 0
+	delimiter := ""
+	output := make(map[string]interface{})
+
+	var step func(object map[string]interface{}, prev string)
+
+	step = func(object map[string]interface{}, prev string) {
+		for key, value := range object {
+			if prev == "" {
+				delimiter = ""
+			} else {
+				delimiter = "."
+			}
+			newKey := prev + delimiter + key
+
+			v, ok := value.(map[string]interface{})
+			shouldStepInside := ok && len(v) > 0 && currentDepth < maxDepth
+			if shouldStepInside {
+				currentDepth++
+				step(v, newKey)
+			} else {
+				output[newKey] = value
+			}
+		}
+	}
+
+	step(target, "")
+	return output
+}
+
+func isStringInSlice(a string, list []string) bool {
+	for _, b := range list {
+		if b == a {
+			return true
+		}
+	}
+	return false
+}
+
+// sortPropNames orders propNames so that timeField is first (if it exists) and rest of propNames are ordered alphabetically
+func sortPropNames(propNames []string, timeField string) []string {
+	sortedPropNames := make([]string, 0, len(propNames))
+	sortedPropNames = append(sortedPropNames, propNames...)
+
+	sort.Strings(sortedPropNames)
+	if len(sortedPropNames) == 0 || sortedPropNames[0] == timeField {
+		return sortedPropNames
+	}
+
+	if sortedPropNames[len(sortedPropNames)-1] == timeField {
+		return append([]string{timeField}, (sortedPropNames)[:len(sortedPropNames)-1]...)
+	}
+
+	for idx, propName := range sortedPropNames {
+		if propName == timeField {
+			return append([]string{timeField}, append((sortedPropNames)[:idx], (sortedPropNames)[idx+1:]...)...)
+		}
+	}
+	return sortedPropNames
+}
+
+// findTheFirstNonNilDocValueForPropName finds the first non-nil value for propName in docs. If none of the values are non-nil, it returns the value of propName in the first doc.
+func findTheFirstNonNilDocValueForPropName(docs []map[string]interface{}, propName string) interface{} {
+	for _, doc := range docs {
+		if doc[propName] != nil {
```

Comment:
We would actually have to do this because  `nil` value is still `ok` value.
```
value, ok := doc[propName]
    if ok && value != nil {
       return value
```

Reply:
I'm going to keep this as it is now, but let me know if you have strong preference and I can update it.

Reply:
no problem, i'm fine with the current approach 👍